docs/guides/high-availability/concepts-memory-and-cpu-sizing.adoc

```diff
@@ -0,0 +1,96 @@
+<#import "/templates/guide.adoc" as tmpl>
+<#import "/templates/links.adoc" as links>
+
+<@tmpl.guide
+title="Concepts for sizing CPU and memory resources"
+summary="Understand these concepts to avoid resource exhaustion and congestion"
+preview="true"
+tileVisible="false" >
+
+Use this as a starting point to size a product environment. 
+Adjust the values for your environment as needed based on your load tests.
+
+== Performance recommendations
+
+[WARNING]
+====
+* Performance will be lowered when scaling to more Pods (due to additional overhead) and using a cross-datacenter setup (due to additional traffic and operations).
+
+* Increased cache sizes can improve the performance when {project_name} instances run for a longer time. Still, those caches need to be filled when an instance is restarted.
+
+* Use these values as a starting point and perform your own load tests before going into production.
+====
+
+Summary:
+
+* The used CPU scales linearly with the number of requests up to the tested limit below.
+* The used memory scales linearly with the number of active sessions up to the tested limit below.
+
+Recommendations:
+
+* The base memory usage for an inactive Pod is 1 GB of RAM.
+
+* Leave 1 GB extra head-room for spikes of RAM.
+
+* For each 100,000 active user sessions, add 500 MB per Pod in a three-node cluster (tested with up to 200,000 sessions).
++
+This assumes that each user connects to only one client.
+Memory requirements increase with the number of client sessions per user session (not tested yet).
+
+* For each 40 user logins per second, 1 vCPU per Pod in a three-node cluster (tested with up to 300 per second).
++
+{project_name} spends most of the CPU time hashing the password provided by the user.
+
+* For each 450 client credential grants per second, 1 vCPU per Pod in a three node cluster (tested with up to 2000 per second).
++
+Most CPU time goes into creating new TLS connections, as each client runs only a single request.
```

Comment:
I also think we should recommend re-encrypting (from ingress) traffic to backend nodes. It is worth mentioning how TLS termination (and the different options like `edge`) can impact this specific usage of CPU?

Reply:
The test succeeded with both sustaining the load and recovering from failures. 

I see this as an optimizations, and maybe you could open an issue in the keycloak benchmark project to run such a test and measure the differences. The Keycloak operator doesn't provide a configuration for certmanager and/or OpenShift to do reencrypt, so once the test for a manually configured Keycloak shows a performance improvement, the next step would be a better Operator support, and then we could eventually update this guide (or the other way around, but it would be a bit tedious for customers)

Reply:
I don't want to put more stuff in `keycloak-benchmark` backlog if not needed. It is more about whether we want to recommend a more secure deployment (which does not really depend on us to decide) with re-encrypt.

Just wanted to highlight this here to see if it makes sense to include it now. Eve if we just want to state the numbers are based on a specific proxy mode.

Reply:
Ok, I'll add a sentence in the follow-up issue https://github.com/keycloak/keycloak/issues/24975